### DISCUSSION TITLE
The Complete Beginner’s Guide to CLI Apps in V | The V Programming Language

### DISCUSSION THREAD

**Discussion by giscus** [vlang/vlang-blog#18] [General]
# The Complete Beginner’s Guide to CLI Apps in V | The V Programming Language

Blog post that shows beginners how easy it is to build a CLI application in V. We build a sample app called geo that prints geometric shapes to the screen.

https://blog.vlang.io/the-complete-beginners-guide-to-cli-apps-in-v/

**Comment by kpeters58** [comment]
Clear & concise - you have a gift for tutorials. Look forward to reading more from you!

**Reply by jkleiser** [reply]
Well written, easy to understand. Thanks!

**Comment by Irwin1985** [comment]
Well written bro, I did follow the tutorial and learn about CLI applications. Thanks!

**Comment by tysonrm** [comment]
Well done.

**Comment by gmlewis** [comment]
Excellent tutorial... thank you!

I'm ran into a couple surprises that maybe you could comment on?

```bash
$ v version
V 0.3.0 f0ce7fb
$ ./geo --version
 
$ ./geo -v
V panic: optional not set (Unknown flag `-v`)
v hash: f0ce7fb
/tmp/v_1000/geo.8635550266121313580.tmp.c:7724: at _v_panic: Backtrace
/tmp/v_1000/geo.8635550266121313580.tmp.c:7687: by panic_optional_not_set
/tmp/v_1000/geo.8635550266121313580.tmp.c:18376: by main__main
/tmp/v_1000/geo.8635550266121313580.tmp.c:18839: by main
```
I expected `-v` to be the short flag version of `--version` and I expected `--version` to print the version number.

Thanks again!

**Reply by hungrybluedev** [reply]
Hey @gmlewis! Glad you found the article to be helpful.

Now about your issue: `-v` is not created as a shorthand for `--version` by default. By older convention, it is used with `--verbose`. If you want to assign the short hand to the version command, you'll have to declare it explicitly for now. Like we did for the other options like `size`. You'll also need to generate the entire help text yourself and store it somewhere. Then show it when the user asks for information.

The `--help` and `--version` tags are conveniently autogenerated when the finalize() function is called. We just had to focus on our custom commands.

**Reply by pepa65** [reply]
I tried to rewrite version like: fp.bool('Version', `V`, false, 'Just display the version')
It didn't work. (My main goal was to be able to use -V as a shorthand, otherwise I would just use .default_version_label)

**Comment by gmlewis** [comment]
Thanks, @hungrybluedev - but I thought `./geo --version` was supposed to give some kind of output, correct?
With V version `0.3.0` I get no output at all.

**Reply by hungrybluedev** [reply]
You're right. I found a bug and reproduced it with a minimal example: https://github.com/vlang/v/issues/15049

As a temporary workaround, the entire manifest instance can be made public and accessed as `geometry.manifest.version`. Then it returns a proper value.

**Reply by hungrybluedev** [reply]
Great news! This should now be fixed with the latest V. 😄

**Comment by kenaryn** [comment]
Hello. Thank you for having wrote this tutorial.

Please why does the `generate_shape()` function in `geo.v ` has un unusual signature with colon (i.e. `:`) between argument's names and its type? This kind of syntax is not documented at all, neither here in the tutorial nor in the official doc for an overview of V.
I thought of a awkwardness or typo at first and omitted the colons but the compiler returned an error.

Can you explain for a newbie the why and how it performs exactly? It would allow me to know when and where using it again in other pieces of code.

**Reply by Irwin1985** [reply]
Hi @kenaryn V refers to it as "Trailing struct literal arguments" and you can check it on it's documentation: https://github.com/vlang/v/blob/master/doc/docs.md#trailing-struct-literal-arguments

**Comment by danawoodman** [comment]
Fun project!

One minor issue with the article is it is using "?" after "finalize" in main.v is deprecated and should be updated to use "!".

V is such an elegant language!

**Comment by nbunnell776** [comment]
I'm receiving compiler error "error: cannot access private field `kind` on `geometry.ShapeOptions`" (repeated for size & symbol) when trying to run either my hand copied code or the version thats on your GH repo... has something changed in the way V accesses public structs in imported modules since this was published?

Otherwise, great tutorial, looking forward to learning more of the language

**Reply by nbunnell776** [reply]
Was reading in the Packt VLang book and discovered that the members also needed to be declared as "pub:" internal to the struct definition, compiled and running now

**Comment by Chiptune11** [comment]
geometry/options.v:12:13: error: struct embedding must be declared at the beginning of the struct body
   10 | 
   11 | pub struct ShapeOptions {
   12 |     pub kind   GeometricShapeKind
      |                ~~~~~~~~~~~~~~~~~~
   13 |     pub size   int
   14 |     pub symbol rune = `*`

**Reply by Chiptune11** [reply]
Ok, that's work:
pub struct ShapeOptions {
	__global:
	kind   GeometricShapeKind
	size   int
	symbol rune = `*`
}
